### ISSUE TITLE
Transformation stopped to work after upgrade

### ISSUE THREAD

**Issue by MaxMorais** [jsonata-js/jsonata#286]
I have the following transformation that works properly into the JSONata version 1.4.1, but into the new versions it doesn't work anymore.

What have changed into JSONata, to it stop to work? And how I can change the transformation for the new versions?

https://gist.github.com/MaxMorais/d641103ff3f0305e9c29d33d9ed08298

**Comment by andrew-coleman** [comment]
It looks like this used to work because of a bug in the `$append()` function that was fixed in `v1.5`.
JSONata is a functional language, not an imperative language.  A consequence of this is that functions cannot have side effects, and all values are immutable.  A [bug was identified](https://github.com/jsonata-js/jsonata/issues/139) in the `$append()` function whereby the first argument was getting modified to have the same array contents as the return value.  Your expression exploits that side effect in a couple of places, which is why it giving a different result since v1.5.  These parts will need re-writing. Feel free to post a message on our slack channel if you want help with that (there's a link to that in the [exerciser](http://try.jsonata.org) toolbar).

I suspect this expression could be simplified in some places.  For example you are using the `$map` and `$filter` functions in places where the `.` and `[]` syntax will suffice.  E.g.
```
$str_item_names := $join($item_names, ",");

$filtered_items := $filter($.items, function($v, $i, $a){
   $contains($str_item_names, $v.item_code)
});
```
could be replaced with
```
$filtered_items := $.items[item_code in $item_names];
```

Closing this issue since it's not a JSONata bug.

**Comment by MaxMorais** [comment]
@andrew-coleman Thank you for your answer!

And glad to know about that bug.
